### DISCUSSION TITLE
changing various settings in init.lua

### DISCUSSION THREAD

**Discussion by etnetu** [orbitalquark/textadept#83] [General]
Hi to all, 
thanks for your work
I'm new to textadept and, but I'm trying to change some settings.

I use a base16 theme but I'd like to change the color of the found word, I used
`view.indic_fore[ui.find.INDIC_FIND] = 0x773300` but no result.

In TeX, I'd like to treat $...$ as braces, I tried
```
    textadept.editing.auto_pairs[string.byte('$')] = '$'
    textadept.editing.typeover_chars[string.byte('$')] = true
    textadept.editing.brace_matches[60] = true 
```
but no result

In TeX I'd like to change the background color of the region between `$ ... $`, and the same for the region between `$$ ... $$`,
is it possible?

many thanks

**Comment by rgieseke** [comment]
This should give you `$` auto pairing for TeX and LaTeX only. I think brace matching works only for distinct braces.

```
events.connect(events.LEXER_LOADED, function(name)
  local is_tex = name == 'tex' or name == 'latex'
  textadept.editing.auto_pairs[string.byte('$')] = is_tex and '$'
  textadept.editing.brace_matches[string.byte('$')] = is_tex
end)
```

**Reply by etnetu** [reply]
thanks for the fast answer,
I tried: now the left $ is highlighted, but the right $ is not

**Reply by rgieseke** [reply]
Sorry, i meant to write
```lua
events.connect(events.LEXER_LOADED, function(name)
  local is_tex = name == 'tex' or name == 'latex'
  textadept.editing.auto_pairs[string.byte('$')] = is_tex and '$'
  textadept.editing.typeover_chars[string.byte('$')] = true
end)
```
I think the brace matching is not possible.

**Reply by etnetu** [reply]
brace matching is defined for " and ' also ...

**Reply by rgieseke** [reply]
Only for auto-pairing, for brace matching only ( [ {

https://github.com/orbitalquark/textadept/blob/default/modules/textadept/editing.lua#L92

**Reply by etnetu** [reply]
ok, thanks
and what about changing the background color, or underline, the region between $...$?

**Comment by rgieseke** [comment]
As for the theming, are you doing this in the theme file or in your `init.lua`?

It needs to be done in the theme file, as this file is loaded for new buffers etc.

There you can also use the defined theme colors, e.g.

```lua
view.indic_fore[ui.find.INDIC_FIND] = colors.base06
view.indic_alpha[ui.find.INDIC_FIND] = 128
```
(I'm maintaining the base16 repo and should probably change the template there, too.)

**Reply by etnetu** [reply]
in the init.lua, and with Ctrl-E

**Comment by rgieseke** [comment]
As for the styling of math environments, you can for example create a copy of `lexers/latex.lua` in a `~/.textadep/lexers` directory and change the styling of the math delimiters:

E.g. change line 28 to
```lua
lex:add_style('math', lexer.styles.embedded)
```
to use the default embedded styling. I'm not sure whether it's possible to change the styling *in between* the delimiters.
Maybe @orbitalquark can add some on this ...

**Reply by rgieseke** [reply]
You can also create your own custom styles or combinations: https://orbitalquark.github.io/textadept/api.html#example-styles

**Reply by etnetu** [reply]
thanks again, the in between is not so important (I tought that was simpler)
as a side note I saw there is tex.lua and latex.lua, if the first is for when the source is not latex (but plain tex for example), how are the two differentiated? so I can load one but not the other

**Reply by rgieseke** [reply]
By default `.tex` is associated with the LaTeX lexer. If one wants to use plain TeX or ConTeXt one could change that (either by changing `textadept.file_types.extensions.tex` or by manually selecting a different lexer from the menu).

**Reply by etnetu** [reply]
ok, many thanks for your patience, you put me on the right tracks

**Comment by orbitalquark** [comment]
Robert is correct in that brace matching/highlighting is only possible for '(', '[', '{', and '<'. That's what Scintilla, the editing component Textadept uses, supports.

As for highlighting in between `$...$`, you would have to create a rule bounded by '$' with a distinct token inside. For example (untested):

    local delim = token('math', P('$'))
    lex:add_rule('math', delim * token('inside_math', (1 - delim)^0) * delim^-1)
    lex:add_style('inside_math', {underline = true})

The 'inside_math' token would get more complicated if you wanted to support escape characters, but you get the idea.
